src/services/request-service.ts

```diff
@@ -0,0 +1,75 @@
+import Promise = require('bluebird');
+import r = require('rethinkdb');
+import uuid = require('uuid');
+import EmailService = require('./email-service');
+import nodemailer = require('nodemailer');
+import q = require('../dbutils/query');
+import models = require('../models/models');
+import errors = require('../errors/errors');
+import config = require('../config');
+
+
+module RequestService {
```

Comment:
So I have been trying a bunch of different possibilities and none of them seem to be free from race conditions.  I tried upserting, inserting with error on conflict, and branching inside the insert so I would hold the write lock. I am open to any ideas, but I am out of them myself.  I really wish rethinkdb just had a unique constraint, that would make life so much easier..